_posts/2019-12-17-zero-to-main-rust-1.md

```diff
@@ -0,0 +1,433 @@
+
+```
+arm-none-eabi-objdump -Ct target/thumbv7em-none-eabihf/release/from-scratch | sort
+...
+00000004 g     O .vector_table  00000004 __RESET_VECTOR
+00000008 g       .vector_table  00000000 __reset_vector
+000000dc g       .vector_table  00000000 _stext
+000000dc l     F .text  0000005c from_scratch::reset_handler
+00000138 l     F .text  00000072 from_scratch::main
+000001ac g       .text  00000000 __etext
+...
+```
+
+Same as in the original post, our compiler has set the lowest bit of our reset handler to one (to indicate a thumb2 function), so `from_scratch::reset_handler` is what we're looking for
+
+## Writing a Reset_Handler
+
+The Cortex-M processor on our board doesn't actually know the difference between C and Rust, so our responsibilities for starting the firmware are the same when we build a Rust program, we need to:
+
+1. Provide an entry point, stored in the second word of flash
+2. Zero-initialize the `.bss` section
+3. Set items with static storage duration to their initial values. In Rust the destination (in RAM) is referred to as `.data`, and the source values (in Flash) is referred to as `.rodata`
+
+Let's go through the code required to write this reset handler, one chunk at a time.
+
+### Starting at the top
+
+The first line in many embedded applications and libraries in Rust will look like this:
+
+```rust
+#![no_std]
+```
+
+This is called a "global attribute", and it is stating that this Rust code will not be using the Rust Standard Library. Attributes in Rust are sometimes used similarly to `#pragma` in C, in that they can change certain behaviors of the compiler.
+
+Rust provides a number of built-in library components, but the two main ones are:
+
+1. The Rust Standard Library
+2. The Rust Core Library
+
+While the Standard Library contains a number of useful components, such as data structures, and interfaces for opening files and sockets, it generally requires your target to have these things! For bare metal applications, we can instead forego this library, and only use the Rust Core Library, does not have these requirements.
+
+Rust as a language has a concept of "modules", which can be used to organize and provide namespaces for your code. Libraries or applications in Rust are called "Crates", and each has it's own namespace. This is why we saw the symbol `from_scratch::reset_handler` in our linker script: It was referring to the `reset_handler` function in the `from_scratch` crate (which is the crate in this example).
+
+To use items from another crate, including the `core` library, you can import these items in using the `use` syntax:
+
+```rust
+use core::{
+    mem::zeroed,
+    panic::PanicInfo,
+    ptr::{read, write_volatile},
+};
+```
+
+This imports the symbols into the current context so that they can be used. Most symbols in Rust are not available in a global namespace, which helps to avoid naming collisions.
```

Comment:
A link to the documentation for the symbols are available in the `core` crate could be useful here.

Reply:
This is more looking at the difference between C, which uses non-namespaced symbols (e.g. items that show up in the linker), vs Rust (which handles namespacing directly).

I could link to the [documentation for the core library](https://doc.rust-lang.org/core/index.html), though this doesn't really directly map to symbols that appear in the final binary (as you could rename or re-export them).

Reply:
Added a little bit more here, let me know if that fits what you were thinking of.

Reply:
👍